google-cloud-spanner/src/test/java/com/google/cloud/spanner/spi/v1/GapicSpannerRpcTest.java

```diff
@@ -502,6 +505,30 @@ public void testAdminRequestsLimitExceededRetryAlgorithm() {
     assertThat(alg.shouldRetry(new Exception("random exception"), null)).isFalse();
   }
 
+  @Test
+  public void testCustomUserAgent() {
+    final HeaderProvider userAgentHeaderProvider =
+        new HeaderProvider() {
+          @Override
+          public Map<String, String> getHeaders() {
+            final Map<String, String> headers = new HashMap<>();
+            headers.put("user-agent", "test-agent");
+            return headers;
+          }
+        };
+    final SpannerOptions options =
+        createSpannerOptions().toBuilder().setHeaderProvider(userAgentHeaderProvider).build();
+    final Spanner spanner = options.getService();
+    final DatabaseClient databaseClient =
+        spanner.getDatabaseClient(DatabaseId.of("[PROJECT]", "[INSTANCE]", "[DATABASE]"));
+    try (final ResultSet rs = databaseClient.singleUse().executeQuery(SELECT1AND2)) {
+      rs.next();
+    }
+
+    assertThat(seenHeaders.get(Key.of("user-agent", Metadata.ASCII_STRING_MARSHALLER)))
```

Comment:
nit: could we add an assertion that the default header provider is also present in the seen headers, so we prevent any silent regression from the current behavior?

Reply:
Good idea, I will add that.